controllers/managedclusterForCGU_controller.go

```diff
@@ -0,0 +1,207 @@
+	reqLogger := r.Log.WithValues("Request.Name", req.Name)
+	reqLogger.Info("Reconciling managedCluster to create clusterGroupUpgrade")
+
+	managedCluster := &clusterv1.ManagedCluster{}
+	err := r.Get(ctx, types.NamespacedName{Name: req.Name}, managedCluster)
+	if errors.IsNotFound(err) {
+		// managed cluster could have been deleted
+		return ctrl.Result{}, nil
+	}
+	if err != nil {
+		// Error reading the object, requeue the request
+		return ctrl.Result{}, err
+	}
+
+	clusterGroupUpgrade := &ranv1alpha1.ClusterGroupUpgrade{}
+	err = r.Get(ctx, types.NamespacedName{Name: req.Name, Namespace: ztpInstallNS}, clusterGroupUpgrade)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			// Error reading clusterGroupUpgrade, requeue the request
+			return ctrl.Result{}, err
+		}
+	} else {
+		// clusterGroupUpgrade for this cluster already exists, stop reconcile
+		r.Log.Info("clusterGroupUpgrade found", "Name", clusterGroupUpgrade.Name, "Namespace", clusterGroupUpgrade.Namespace)
+		return ctrl.Result{}, nil
+	}
+
+	availableCondition := meta.FindStatusCondition(managedCluster.Status.Conditions, clusterv1.ManagedClusterConditionAvailable)
+	if availableCondition == nil {
+		r.Log.Info("cluster has no status yet", "Name", managedCluster.Name, "RequeueAfter:", clusterNoStatusRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterNoStatusRetryDelay}, nil
+	} else if availableCondition.Status == metav1.ConditionTrue {
+		// cluster is ready
+		policies, err := r.getPolicies(ctx, managedCluster.Name)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if policies == nil {
+			// child policies have not been created, requeue the request
+			return ctrl.Result{RequeueAfter: clusterNotReadyRetryDelay}, nil
+		}
+
+		// create clusterGroupUpgrade
+		if err = r.newClusterGroupUpgrade(ctx, managedCluster, policies); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		// availableCondition is false or unknown, cluster is not ready
+		r.Log.Info("cluster is not ready", "RequeueAfter:", clusterNotReadyRetryDelay)
+		return ctrl.Result{RequeueAfter: clusterNotReadyRetryDelay}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ManagedClusterForCguReconciler) getPolicies(ctx context.Context, clusterName string) (*policiesv1.PolicyList, error) {
+	policies := &policiesv1.PolicyList{}
+	if err := r.List(ctx, policies, client.InNamespace(clusterName)); err != nil {
+		return nil, err
+	}
+	return policies, nil
+}
+
+// Create a clusterGroupUpgrade
+func (r *ManagedClusterForCguReconciler) newClusterGroupUpgrade(ctx context.Context, cluster *clusterv1.ManagedCluster, policies *policiesv1.PolicyList) (err error) {
+	var managedPolicies []string
+	var policyWaveMap = make(map[string]string)
+	for _, policy := range policies.Items {
+		deployWave := policy.GetAnnotations()[ztpDeployWaveAnnotation]
+		// apply the policy at the last if policy doesn't have wave
+		if deployWave == "" {
+			deployWave = "1000"
```

Comment:
Could we add a comment or link and explain what there wave values means?

Reply:
Will do